File: Python_version/stack/stack.py




class Stack:
    def __init__(self, max_size):
        self.max_size = max_size
        self.stack = [0] * max_size
        self.top = -1

    def is_empty(self):
        if self.top == -1:
            return True
        else:
            return False

    def push(self, data):
        if self.top >= self.max_size - 1:
            print('堆疊已滿')
        else:
            self.top += 1
            self.stack[self.top] = data

    def pop(self):
        if self.is_empty():
            print('堆疊已空')
        else:
            print('彈出 %d' % self.stack[self.top])
            self.top -= 1

    def print_stack(self):
        for i in range(self.top, -1, -1):
            print('--%d--' % self.stack[i])
        print('-----')

if __name__ == '__main__':
    max_size = 5
    s = Stack(max_size)
    s.push(1)
    s.push(2)
    s.push(3)
    s.pop()
    s.print_stack()

File: C_version/stack/stack.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
/* stack project */

#define NUM 100
int stack[NUM];
int top = -1;
// 使用三個函式(放入push、彈出pop、列印print)
void push(int);
int pop(void);
void printstack(void);

// 主函式
int main(void){
    int choice =0,n;
    while (choice != -1){  // 設立判斷式 只要不為 -1 就不跳出迴圈
        printf("1) push 2) pop 3) printstack  -1) quit\n");
        printf("enter your chioce: ");
        scanf("%d",&choice);
        switch(choice){  //使用 switch 進行功能選擇
            case 1:
                printf("what do you want to input:\n");
                scanf("%d",&n);
                push(n);
                break;
            case 2:
                printf("%d was popped from stack \n",pop());
                break;
            case 3:
                printstack();
                break;
            
        }
    }
}

void push(int n){
    if (top == NUM-1){
        printf("stack is fulled");
        exit(1); //堆疊滿了，exit(0) 表示程式正常結束，而 exit(1) 則表示程式異常結束
    }
    top++;
    stack[top] = n;
}

int pop(void){
    if(top ==-1){
        printf("stack is empty\n");
        exit(1); //堆疊資料空了 跳出程式，exit(0) 表示程式正常結束，而 exit(1) 則表示程式異常結束
        
    }
    top--;
    return stack[top];
}

void printstack(void){
    int i;
    if(top ==-1){
        printf("stack is empty:\n");
    }else{
        printf("stack\n");
        printf("-------------\n");
        for (i=top;i>=0;i--){
            printf("%d",stack[i]);
        printf("\n");
        }
    }
}

